src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,284 @@
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Handles POST and GET requests for projects,
+ * allows creation and update of projects, and
+ * supports queries for projects based on various parameters.
+ */
+@WebServlet("/projects")
+public class ProjectServlet extends HttpServlet {
+
+  /**
+   * Handles POST requests for projects.
+   * Responds with project ID upon successful POST.
+   * @param     {HttpServletRequest}    request
+   * @param     {HttpServletResponse}   response
+   * @return    {void}
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Must be logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/"); // placeholder: should redirect to login
+      return;
+    }
+
+    // Mode is a required parameter
+    boolean isCreateMode = DataUtils.parseMode(request, response);
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+
+    // Will be either a new entity for creation or an existing entity for
+    // updating
+    Entity projEntity = new Entity(DataUtils.PROJECT);
+
+    if (!isCreateMode) {
+      // Must be owner to update
+      projEntity = DataUtils.getProjectEntity(projId, uEmail, false, false);
+
+      // Delete overrides all other updates
+      boolean delete = Boolean.parseBoolean(request.getParameter("delete"));
+      if (delete) {
+        datastore.delete(projEntity.getKey());
+        response.sendRedirect("/"); // placeholder: should redirect to projects gallery
+        return;
+      }
+    }
+
+    // Last modified timestamp
+    String now = Instant.now().toString();
+    projEntity.setProperty("utc", now);
+
+    String projName = request.getParameter("proj-name");
+
+    // Set the project name if provided
+    // If creating and nothing provided, set name to Untitled-{current UTC time}
+    if (!DataUtils.isEmptyParameter(projName)) {
+      projEntity.setProperty("name", projName);
+    } else {
+      if (isCreateMode) {
+        projEntity.setProperty("name", "Untitled-" + now);
+      }
+      // Don't do anything if updating and no name provided
+    }
+
+    // Default visibility to private (only owners and editors can view)
+    String visibility = request.getParameter("visibility");
+    if (DataUtils.isEmptyParameter(visibility) && isCreateMode) {
+      visibility = DataUtils.PRIVATE;
+    }
+    if (!DataUtils.isEmptyParameter(visibility) &&
+        (visibility.toLowerCase().equals(DataUtils.PUBLIC) ||
+         visibility.toLowerCase().equals(DataUtils.PRIVATE))) {
+      projEntity.setProperty("visibility", visibility);
```

Comment:
Might be better to store `visibility.toLowerCase()`, so that the values in your database are standardized.